### PR TITLE
feat(mcp): add vault_patch_note and vault_replace_in_note tools

### PR DESCRIPTION
**Pull request by aliasunder**
## Summary

- **`vault_patch_note`** — surgical heading-targeted edits with 4 operations (append, prepend, replace, insert_before). Supports file-level append/prepend (no heading), CommonMark §4.5 code-fence-aware H1–H6 parser (backticks + tildes, nested fences), case-sensitive heading matching with ambiguity detection (same-level vs cross-level error messages), and frontmatter preservation.
- **`vault_replace_in_note`** — exact text find-and-replace with optional `replace_all_occurrences` flag. Empty `old_text` rejected at schema and runtime. Case-sensitive, frontmatter preserved.
- New `vault-patcher.ts` module with `applySectionOperation` helper, immutable data flow (no `let`), explicit naming throughout
- `resolveSafePath` exported from `vault-filesystem.ts` for reuse
- Tool descriptions include `Errors:` sections with remediation guidance, following MCP best practices
- 53 new tests (251 total), 15 registered tools

### Doc updates
- ARCHITECTURE.md: added `vault_patch_note`, `vault_replace_in_note`, `vault_search_by_folder` to tool tables; fixed stale `sort_by` value (`mtime` → `modified`)
- AGENTS.md: added conventions (immutability, explicit naming, regex doc comments, MCP tool description template); updated tool count and data-layer module list
- README.md: updated tool count, removed stale test count

### Ultrareview findings addressed
- Empty `oldText` silently corrupts note → runtime guard + `.min(1)` schema
- `idempotentHint` wrong on `vault_replace_in_note` → `false`
- Misleading example targeted frontmatter (body-only tool) → body-resident example
- `heading_level` accepted floats → added `.int()`
- Nested code fences mishandled → CommonMark §4.5 fence tracking
- Same-level ambiguity error suggested `heading_level` (unhelpful) → suggests rename
- Missing `Returns:` lines → added to both tools

## Test plan

- [x] 53 new tests: heading parsing (6 levels, code blocks, nested fences, tilde fences, trailing hashes), heading lookup (case sensitivity, ambiguity with same-level vs cross-level messages, disambiguation, empty heading), all 4 patchNote operations (file-level and section-level with structural position verification), frontmatter preservation (with/without/complex, list-type values), replaceInNote (first/all/multi-line/deletion/case-sensitive/empty oldText), error cases (not found, path traversal), edge cases (empty sections, trailing headings, subdirectories)
- [x] All 251 tests pass
- [x] TypeScript compiles clean
- [x] ESLint passes
- [x] CI passing on all pushes
- [x] Deploy and verify tools via MCP Inspector or Claude Desktop

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/vault-mcp/vault-patcher.ts [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+  bodyEndLine: number
+}>
+
+// ── Internal helpers ────────────────────────────────────────────
+
+const HEADING_RE = /^(#{1,6}) (.+)$/
+
+/**
+ * Single-pass heading parser for H1–H6 with code-block awareness.
+ * Section body = heading+1 through next heading of same-or-higher level (or EOF).
```
Comment (aliasunder) [inline src/vault-mcp/vault-patcher.ts:32]: Should this be above the function vs this var?

File: src/vault-mcp/vault-patcher.ts [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+  text: string
+  level: number
+  startLine: number
+  bodyStartLine: number
+  bodyEndLine: number
+}>
+
+// ── Internal helpers ────────────────────────────────────────────
+
+const HEADING_RE = /^(#{1,6}) (.+)$/
```
Comment (aliasunder) [inline src/vault-mcp/vault-patcher.ts]: What does "re" stand for. It's better to be explicit in name instead of using vague short forms that may not be understood.

File: src/vault-mcp/vault-patcher.ts [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+
+/**
+ * Single-pass heading parser for H1–H6 with code-block awareness.
+ * Section body = heading+1 through next heading of same-or-higher level (or EOF).
+ */
+const FENCE_OPEN_RE = /^(`{3,}|~{3,})/
```
Comment (aliasunder) [inline src/vault-mcp/vault-patcher.ts]: Same issue here as other const. pthis const and the other should have a doc string providing a short explanation on what it does. Regex is not super obvious at first glance, so should always include a doc string/ comment

File: src/vault-mcp/vault-patcher.ts [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+
+const parseHeadings = (lines: readonly string[]): HeadingInfo[] => {
+  let fence: { char: string; length: number } | null = null
```
Comment (aliasunder) [inline src/vault-mcp/vault-patcher.ts]: Any way we can make the logic in here avoid object mutation ? Let me know your thoughts.

File: src/vault-mcp/vault-patcher.ts [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+      acc.push({
+        text: match[2].replace(/\s+#+\s*$/, "").trim(),
+        level: match[1].length,
+        startLine: i,
+      })
+    }
+    return acc
+  }, [])
+
+  return raw.map((h, i) => {
```
Comment (aliasunder) [inline src/vault-mcp/vault-patcher.ts]: This logic could really use some comments as the logic is a bit more complex and less readable than a more simple function. Any logic that is not super obvious should have comments. Do a pass of all changes to identify where a comment or doc string would be useful to a person reading the code.

File: src/vault-mcp/vault-patcher.ts [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+const findHeading = (
+  headings: readonly HeadingInfo[],
+  text: string,
+  level?: number,
+): HeadingInfo => {
+  if (!text.trim()) {
+    throw new Error("heading cannot be empty")
+  }
+
+  const needle = text.trim()
```
Comment (aliasunder) [inline src/vault-mcp/vault-patcher.ts]: Needle? What's that?

File: src/vault-mcp/vault-patcher.ts [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+const readNoteForPatch = async (
+  vaultPath: string,
+  path: string,
+): Promise<{
+  fullPath: string
+  data: Record<string, unknown>
+  lines: string[]
+}> => {
+  const fullPath = resolveSafePath(vaultPath, path)
+  let raw: string
```
Comment (aliasunder) [inline src/vault-mcp/vault-patcher.ts]: Any way to is can be updated to be immutable?

File: src/vault-mcp/vault-patcher.ts [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+  logger: Logger,
+): Promise<string> => {
+  const { path, operation, content, heading, headingLevel } = params
+  const { fullPath, data, lines } = await readNoteForPatch(
+    params.vaultPath,
+    path,
+  )
+  const contentLines = content.split("\n")
+
+  let updatedLines: string[]
```
Comment (aliasunder) [inline src/vault-mcp/vault-patcher.ts]: Do a pass of all logic for opportunities to make the code/logic use less object mutation. 

Now if making it immutable makes the code less clear or more complex, then we can leave it, but we should be leaning towards immutability wherever it makes sense

File: src/vault-mcp/vault-patcher.ts [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+    throw new Error("heading cannot be empty")
+  }
+
+  const searchText = text.trim()
+  const matches = headings.filter(
+    (h) => h.text === searchText && (level === undefined || h.level === level),
+  )
+
+  if (matches.length === 0) {
+    const available = headings
```
Comment (aliasunder) [inline src/vault-mcp/vault-patcher.ts]: Available? Can this bar be named more explicitly ?

File: src/vault-mcp/vault-patcher.ts [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+
+  const { fullPath, data, lines } = await readNoteForPatch(
+    params.vaultPath,
+    path,
+  )
+
+  const body = lines.join("\n")
+
+  if (!body.includes(oldText)) {
+    const preview = oldText.length > 80 ? oldText.slice(0, 80) + "…" : oldText
```
Comment (aliasunder) [inline src/vault-mcp/vault-patcher.ts]: Preview? Again can be named more explicitly

File: src/vault-mcp/tool-definitions.ts [hunk 11]
```diff
@@ -171,6 +174,139 @@ Returns: Confirmation message.`,
...
+
+Operations:
+- append: add content at end of section (or end of file if no heading)
+- prepend: add content after heading line (or after frontmatter if no heading)
+- replace: replace section body (heading preserved; requires heading)
+- insert_before: insert content above the heading line (requires heading)
+
+Section boundaries: a section spans from its heading to the next heading of the same or higher level (or EOF). Child headings are included in the parent section.
+
+Returns: Confirmation message.`,
```
Comment (aliasunder) [inline src/vault-mcp/tool-definitions.ts:202]: Would it be useful to include the possible errors that could be returned in the description? And are we following best practices there for errors? Use MCP-builder skill to assist

File: src/vault-mcp/vault-patcher.ts [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+const readNoteForPatch = async (
+  vaultPath: string,
+  path: string,
+): Promise<{
+  fullPath: string
+  data: Record<string, unknown>
+  lines: string[]
+}> => {
+  const fullPath = resolveSafePath(vaultPath, path)
+  const raw = await readFile(fullPath, "utf8").catch((err) => {
```
Comment (aliasunder) [inline src/vault-mcp/vault-patcher.ts]: Raw what - also async / await is preferable. A bit of duplication is ok for this to be immutable and use async/await

File: src/vault-mcp/vault-patcher.ts [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+    const availableHeadings = headings
+      .map((h) => `${"#".repeat(h.level)} ${h.text}`)
+      .join(", ")
+    throw new Error(
+      `heading not found: "${searchText}". Available headings: ${availableHeadings || "(none)"}`,
+    )
+  }
+
+  if (matches.length > 1) {
+    const details = matches
```
Comment (aliasunder) [inline src/vault-mcp/vault-patcher.ts]: Could be more explicit

File: src/vault-mcp/tool-definitions.ts [hunk 14]
```diff
@@ -171,6 +174,150 @@ Returns: Confirmation message.`,
...
+Operations:
+- append: add content at end of section (or end of file if no heading)
+- prepend: add content after heading line (or after frontmatter if no heading)
+- replace: replace section body (heading preserved; requires heading)
+- insert_before: insert content above the heading line (requires heading)
+
+Section boundaries: a section spans from its heading to the next heading of the same or higher level (or EOF). Child headings are included in the parent section.
+
+Errors:
+- "note not found" — path does not exist; check vault_list_notes for valid paths
```
Comment (aliasunder) [inline src/vault-mcp/tool-definitions.ts:197]: Nice descriptions. Does the return format of these errors follow MCP tool best practices?

File: src/vault-mcp/vault-patcher.ts [hunk 15]
```diff
@@ -0,0 +1,314 @@
...
+  } catch (err) {
+    if ((err as NodeJS.ErrnoException).code === "ENOENT") {
+      throw new Error(`note not found: "${path}"`, { cause: err })
+    }
+    throw err
+  }
+}
+
+/** Writes modified content back with preserved frontmatter. */
+const writePatched = async (
```
Comment (aliasunder) [inline src/vault-mcp/vault-patcher.ts]: Can't this be named 'writePatchedFile'? Less ambiguous
